### PR TITLE
asio strand scheduler

### PR DESCRIPTION
**Pull request by CorentinBT**
## Summary by CodeRabbit

- **New Features**
  - Introduced a new build configuration option to enable ASIO support in the RPP library.
  - Added a new `strand` class for thread-safe task scheduling using ASIO's strand functionality.
  - Integrated the `worker_strategy` class to facilitate task execution strategies within the `strand` class.

- **Documentation**
  - Updated documentation to clarify the integration of ASIO with ReactivePlusPlus.

- **Tests**
  - Added unit tests for the strand scheduler to validate its functionality and behavior in various scenarios.

### CODE REVIEW

File: src/extensions/rppasio/rppasio/schedulers/strand.hpp [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+#include <rpp/schedulers/details/worker.hpp>
+
+#include <asio/basic_waitable_timer.hpp>
+#include <asio/bind_executor.hpp>
+#include <asio/strand.hpp>
+
+namespace rppasio::schedulers
+{
+    /**
+     * @brief TODO
```
Comment (victimsnino) [inline src/extensions/rppasio/rppasio/schedulers/strand.hpp]: TODO =)

**Comment by victimsnino** [comment]
@CorentinBT , sonarcloud showing valid issues:
https://sonarcloud.io/project/issues?severities=BLOCKER%2CCRITICAL%2CMAJOR%2CMINOR&sinceLeakPeriod=true&issueStatuses=OPEN%2CCONFIRMED&types=BUG&pullRequest=627&id=victimsnino_ReactivePlusPlus&open=AZGlXtnEPhZvl699Zupo

you are trying to forward values but they just forwarded inside lambda: from this point it is regular variable and should be just moved. Forward inside lambda is ok in case of capturing by reference

**Comment by CorentinBT** [comment]
> @CorentinBT , sonarcloud showing valid issues: https://sonarcloud.io/project/issues?severities=BLOCKER%2CCRITICAL%2CMAJOR%2CMINOR&sinceLeakPeriod=true&issueStatuses=OPEN%2CCONFIRMED&types=BUG&pullRequest=627&id=victimsnino_ReactivePlusPlus&open=AZGlXtnEPhZvl699Zupo
> 
> you are trying to forward values but they just forwarded inside lambda: from this point it is regular variable and should be just moved. Forward inside lambda is ok in case of capturing by reference

done

File: src/extensions/rppasio/rppasio/schedulers/strand.hpp [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+                        const auto top = m_queue.pop();
+                        if (top->is_disposed())
+                            continue;
+
+                        m_state.defer_to(
+                            top->get_timepoint(),
+                            [top](const auto&) -> rpp::schedulers::optional_delay_to {
+                                if (const auto advanced_call = top->make_advanced_call())
+                                {
+                                    top->set_timepoint(worker_strategy::now());
```
Comment (CorentinBT) [inline src/extensions/rppasio/rppasio/schedulers/strand.hpp]: @victimsnino a bit uncertain if I should set `now()` here

Comment (victimsnino) [reply src/extensions/rppasio/rppasio/schedulers/strand.hpp]: UPD: you should set to `top->handle_advanced_call(*advanced_call)` here

it is how queue works https://github.com/victimsnino/ReactivePlusPlus/blob/v2/src/rpp/rpp/schedulers/details/queue.hpp#L229

Comment (CorentinBT) [reply src/extensions/rppasio/rppasio/schedulers/strand.hpp]: but then what do I return as `delay_to`?

Comment (victimsnino) [reply src/extensions/rppasio/rppasio/schedulers/strand.hpp]: exactly the same

File: src/extensions/rppasio/rppasio/schedulers/strand.hpp [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+                    timer->async_wait(asio::bind_executor(m_strand, [self = this->shared_from_this(), timer, fn = std::forward<Fn>(fn), handler = std::forward<Handler>(handler), ... args = std::forward<Args>(args)](const asio::error_code& ec) mutable {
+                        if (ec || handler.is_disposed())
+                            return;
+
+                        current_thread_queue_guard guard{*self};
+                        if (const auto new_duration = fn(handler, args...))
+                            self->defer_for(new_duration->value, std::move(fn), std::move(handler), std::move(args)...);
+                    }));
+                }
+            }
```
[the remark this answers is not in the document]
Comment (victimsnino) [reply src/extensions/rppasio/rppasio/schedulers/strand.hpp]: but it is correct usage

File: src/extensions/rppasio/rppasio/schedulers/strand.hpp (hunk 2 above)
Comment (CorentinBT) [reply src/extensions/rppasio/rppasio/schedulers/strand.hpp]: done

**Pull request merged**